src/main/scala/util/Annotations.scala

```diff
@@ -3,12 +3,98 @@
 package freechips.rocketchip.util
 
 import Chisel._
+import chisel3.internal.InstanceId
 import chisel3.experimental.{annotate, ChiselAnnotation, RawModule}
 import firrtl.annotations._
+import freechips.rocketchip.diplomacy._
+import freechips.rocketchip.tilelink.TLToAXI4IdMapEntry
 
-// TODO: replace this with an implicit class when @chisel unprotects dontTouchPorts
-trait DontTouch { self: RawModule =>
+/** Record a case class that was used to parameterize this target. */
+case class ParamsAnnotation(target: Named, paramsClassName: String, params: Map[String,Any]) extends SingleTargetAnnotation[Named] {
+  def duplicate(n: Named) = this.copy(n)
+}
+
+case class ParamsChiselAnnotation[T <: Product](target: InstanceId, params: T) extends ChiselAnnotation {
+  private val paramMap = params.getClass.getDeclaredFields.map(_.getName).zip(params.productIterator.to).toMap
```

Comment:
As far as I know, there isn't a better way to type bound case classes than by using `Product`, but note that lots of things implement `Product`, including `List`:
```scala
scala> val xs = List(1, 2, 3)
xs: List[Int] = List(1, 2, 3)

scala> xs.getClass.getDeclaredFields.map(_.getName).zip(xs.productIterator.to).toMap
res1: scala.collection.immutable.Map[String,Any] = Map(serialVersionUID -> 1, head -> List(2, 3))
```
The even more hilarious thing about the `List` above, is that 1 isn't actually the `serialVersionUID` of the `List`, this trick for case classes corresponds the names and values incorrectly for `List`.

Reply:
Yeesh, I see, that is an annoying hole. Do you know any way to narrow down the scope further, e.g. I know cases classes also extend `Serializable`? Do Lists?

Reply:
Yeah Lists also extend `Serializable`, I'm not sure if there's much more we can do here.